File: reacjilator-java-jslack/src/main/java/slackapp_backend/handler/EchoHandler.java
package slackapp_backend.handler;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.github.seratch.jslack.app_backend.vendor.aws.lambda.request.ApiGatewayRequest;
import com.github.seratch.jslack.app_backend.vendor.aws.lambda.response.ApiGatewayResponse;
import com.github.seratch.jslack.app_backend.vendor.aws.lambda.util.SlackSignatureVerifier;
import lombok.extern.slf4j.Slf4j;
import slackapp_backend.service.AmazonWebServices;

import java.util.Collections;

// This is just a minimal working example
@Slf4j
public class EchoHandler implements RequestHandler<ApiGatewayRequest, ApiGatewayResponse> {

    private final SlackSignatureVerifier signatureVerifier = new SlackSignatureVerifier();
    private final AmazonWebServices aws = new AmazonWebServices();

    @Override
    public ApiGatewayResponse handleRequest(ApiGatewayRequest req, Context context) {
        log.info("request: {}", req);

        if (!signatureVerifier.isValid(req) && !aws.isLocalDev(context)) {
            return ApiGatewayResponse.builder().statusCode(401).build();
        }

        String body = req.getBody();
        if (body != null && body.equals(WarmupHandler.PAYLOAD_STRING)) {
            return ApiGatewayResponse.builder().statusCode(200).build();
        } else {
            return ApiGatewayResponse.builder()
                    .statusCode(200)
                    .headers(Collections.singletonMap("Content-Type", "application/json"))
                    .objectBody(req.getQueryStringParameters())
                    .build();
        }
    }

}

File: reacjilator-scala-jslack/src/main/scala/slackapp/handler/EchoHandler.scala
package slackapp.handler

import com.amazonaws.services.lambda.runtime.{ Context, RequestHandler }
import com.github.seratch.jslack.app_backend.vendor.aws.lambda.request.ApiGatewayRequest
import com.github.seratch.jslack.app_backend.vendor.aws.lambda.response.ApiGatewayResponse
import com.github.seratch.jslack.app_backend.vendor.aws.lambda.util.SlackSignatureVerifier
import org.slf4j.LoggerFactory
import slackapp.service.AmazonWebServices

import scala.collection.JavaConverters._

// This is just a minimal working example
class EchoHandler extends RequestHandler[ApiGatewayRequest, ApiGatewayResponse] {
  import EchoHandler._

  override def handleRequest(req: ApiGatewayRequest, context: Context): ApiGatewayResponse = {
    log.info(s"request: ${req}")

    if (!signatureVerifier.isValid(req) && !aws.isLocalDev(context)) {
      response.statusCode(401).build
    } else {
      req.getBody match {
        case WarmupHandler.payloadString =>
          response.statusCode(200).build
        case _ =>
          response
            .statusCode(200)
            .headers(Map("Content-Type" -> "application/json").asJava)
            .objectBody(req.getQueryStringParameters)
            .build
      }
    }
  }

}

object EchoHandler {

  private val log               = LoggerFactory.getLogger(classOf[EchoHandler])
  private val signatureVerifier = new SlackSignatureVerifier
  private val aws               = new AmazonWebServices

  private def response = ApiGatewayResponse.builder

}
